src/translate.py

```diff
@@ -4241,207 +4221,34 @@ def current_instr(self, instr: Instruction) -> Iterator[None]:
             self.in_pattern = False
 
 
-def process_instruction(instr: Instruction, state: NodeState) -> None:
-    stack_info = state.stack_info
-    arch = stack_info.global_info.arch
-    mnemonic = instr.mnemonic
-    arch_mnemonic = instr.arch_mnemonic(arch)
-    args = InstrArgs(instr.args, state.regs, stack_info)
-    expr: Expression
-
-    # Figure out what code to generate!
-    if mnemonic in arch.instrs_ignore:
-        pass
-
-    elif mnemonic in arch.instrs_store or mnemonic in arch.instrs_store_update:
-        # Store a value in a permanent place.
-        if mnemonic in arch.instrs_store:
-            to_store = arch.instrs_store[mnemonic](args)
-        else:
-            # PPC specific store-and-update instructions
-            # `stwu r3, 8(r4)` is equivalent to `$r3 = *($r4 + 8); $r4 += 8;`
-            to_store = arch.instrs_store_update[mnemonic](args)
-
-            # Update the register in the second argument
-            update = args.memory_ref(1)
-            if not isinstance(update, AddressMode):
-                raise DecompFailure(
-                    f"Unhandled store-and-update arg in {instr}: {update!r}"
-                )
-            state.set_reg(
-                update.rhs,
-                add_imm(args.regs[update.rhs], Literal(update.offset), stack_info),
-            )
-
-        # `to_store` is None for preserving registers in function preludes (which are elided)
-        if to_store is not None:
-            state.store_memory(
-                source=to_store.source, dest=to_store.dest, reg=args.reg_ref(0)
+def evaluate_instruction(instr: Instruction, state: NodeState) -> None:
+    # Check that instr's attributes are consistent
+    if instr.is_return:
+        assert isinstance(state.node, ReturnNode)
+    if instr.is_conditional:
+        assert state.branch_condition is None and state.switch_control is None
+
+    # Handle function_targets here, instead of inside eval_fn, because we all the
+    # information we need is already encoded in the Instruction
+    if instr.function_target is not None:
```

Comment:
Function calls need a comment in arch_* if we are to leave this here. Weakly in favor of moving this logic into arch_* though, to have a single place where we handle instructions.

(Also pondering whether we can make `function_target` a boolean property, but then return type narrowing would need to poke at raw instruction arguments, so probably a bad idea)